iterative/dvc.org
# scroll sidebar follows along with main scroll

## Problem
This PR tackles https://github.com/iterative/dvc.org/issues/1333. The problem is explained in [this video.](https://share.vidyard.com/watch/eGwnBPBjWmyAF2Akp7BBmb)
Basically the right panel of anchor tags doesn't scroll with the page. Thus when readers scroll down to the later parts of the page, the right panel fails to provide value in showing them where they are in the anchor tag table of contents.

<img width="1359" alt="DVC_Core_Project___Data_Version_Control_·_DVC" src="https://user-images.githubusercontent.com/22982964/124348281-9c555d80-dc1b-11eb-8ca4-ade3afbc43b9.png">


## Solution
We force the right sidebar to scroll with the page so that readers always know where they are. 
At first I used a `<HTMLDivElement>.scrollIntoView()` based solution, but that was not smooth in Firefox because it affected the whole screen. I switched to a `<HTMLDivElement>.scrollTop` based solution, which works smoothly in Chrome/Firefox/Safari. 

[Video Demo](https://share.vidyard.com/watch/fNKZuv7hzco8Ki8kgfh8jJ?)

![2021-07-03 15 58 15](https://user-images.githubusercontent.com/22982964/124347457-86459e00-dc17-11eb-90ea-9b7762d2c267.gif)

## Notes:
- The scroll function is debounced for kinder browser performance. You can further increase the debounce duration, but this comes at the cost of the scroll bar appearing to "lag" slightly. The current 100ms is fine as the page is mostly static and isn't scare of processing resources
- On Safari, the very top of the page will not highlight the `How to report a problem` link, which is also the previous behavior in production. Do we want to push further and get that minor part working perfectly too? How does that fit into our triaged list of priorities?
- On Chrome/Firefox/Safari, if you scroll up very very fast, the right panel scrollbar won't keep up. However, you have to try quite hard to make this bug appear. For the majority of users they will not see this. Do we want to push further and get that minor part working perfectly too? How does that fit into our triaged list of priorities?
- Total time, 1.5 hours

## Testing Instructions
0. We will be replicating [these test steps](https://share.vidyard.com/watch/fNKZuv7hzco8Ki8kgfh8jJ?)
1. Visit localhost link http://localhost:8000/doc/user-guide/contributing/core
2. Resize the page shorter to force a smaller right panel (this gives us more scroll bar to see in action)
3. Scroll down the page and watch the right panel scroll with you to show the latest highlighted link
4. Scroll up the page to see the same behavior going upwards
5. Repeat this test for Chrome, Firefox & Safari. This only needs to be done in Desktop as the right panel only appear in desktop width screens
